lib/core/src/Cardano/Wallet/Primitive/Types/UTxOIndex.hs

```diff
@@ -0,0 +1,55 @@
+-- |
+-- Copyright: © 2018-2021 IOHK
+-- License: Apache-2.0
+--
+-- Provides the 'UTxOIndex' type, which indexes a UTxO set by asset ID.
+--
+-- The index makes it possible to efficiently compute the subset of a UTxO
+-- set containing a particular asset, or to select a UTxO entry containing
+-- that asset, without having to search through the entire UTxO set.
```

Comment:
I have some reservations about this entire module. The claim seems to be that computing a subset of a UTxO containing a particular asset id would be too slow, yet.. is that really the case?

Partitioning a UTxO can be done in one pass, and is only needed before running a coin-selection. Even for large UTxO sets this should already be pretty fast to do, and it's a cost we would pay only for coin selections. Now, we'd be paying the cost of the UTxOIndex for every single block application. Although it may be small, it's also impacting a much larger part of the code for something that is only truly needed at one (infrequent) point in time 

Wouldn't it be easier to "flatten" the `TokenMap` instead to facilitate the partitioning, and do the partitioning when setting up the context for coin selections:

```hs
newtype TokenMap = TokenMap
    { unTokenMap
        :: Map AssetId TokenQuantity
    }
```

```hs
partitionUTxO :: UTxO -> NonEmpty AssetId -> Map AssetId UTxO
```

Reply:
> I have some reservations about this entire module. The claim seems to be that computing a subset of a UTxO containing a particular asset id would be too slow, yet.. is that really the case?
> Partitioning a UTxO can be done in one pass, and is only needed before running a coin-selection. Even for large UTxO sets this should already be pretty fast to do, and it's a cost we would pay only for coin selections. Now, we'd be paying the cost of the UTxOIndex for every single block application. Although it may be small, it's also impacting a much larger part of the code for something that is only truly needed at one (infrequent) point in time

It seems we're comparing:

- **incremental indexing**: where we update the index in small chunks, as we parse blocks.
- **just-in-time indexing**: where we build the index in one go, before each coin selection.

If we believe that "just-in-time indexing" is acceptable for now, then we could simply remove the following commit from this PR, which integrates `UTxOIndex` into the `Wallet` type:

https://github.com/input-output-hk/cardano-wallet/pull/2431/commits/43bbeecd580dd0d6952223cbc815ec6c380884e6

This of course leaves us with the option of re-integrating it later on, if we discover (through benchmarking) that incremental indexing is preferable (perhaps for users of wallets with very large UTxO sets).

> Wouldn't it be easier to "flatten" the TokenMap instead to facilitate the partitioning, and do the partitioning when setting up the context for coin selections:
> ```
> newtype TokenMap = TokenMap
>    { unTokenMap
>        :: Map AssetId TokenQuantity
>    }
> partitionUTxO :: UTxO -> NonEmpty AssetId -> Map AssetId UTxO
> ```

This is exactly the type of feedback that would have been useful at the design stage.

Unfortunately, we're now very close to the deadline, and I suspect there is no time budget for making fundamental alterations to the design.

There are course many different ways to slice this problem.

When selecting a UTxO entry at random that contains a given asset, if we choose to select that entry, then we also need to remove it from consideration by future selections. In your proposed simplification, we'd need to remove the entry from all the values of the map returned by `partitionUTxO`.

But this is essentially what the `UTxOIndex` type (and `selectRandom` operation) already does. In addition, it also keeps track of the current balance, which is something we need to keep track of when running coin selection (both for the selected and remaining UTxO sets).

Given that we're under a tight deadline, and given that the existing coin selection code is built with the `UTxOIndex` type in mind, and given that it is fully tested and already been through the review process, I propose that we:

1. Keep the existing implementation of `UTxOIndex`.
2. Remove the commit integrating it into the `Wallet` type.
3. Create the index only when we need to run a coin selection.

Thoughts?

Reply:
> Thoughts?

Yes